### PR TITLE
feat: Add SDK for External Volumes

### PR DESCRIPTION
**Pull request by jdoldis**
This PR adds the SDK for External Volumes. This is part of a series of PRs aimed at adding provider support for Iceberg tables.

## References
- https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2980
- https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2249

### CODE REVIEW

File: pkg/sdk/testint/external_volumes_gen_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+			},
+		)
+	})
+
+	t.Run("Describe", func(t *testing.T) {
+		allowWrites := true
+		comment := "some comment"
+		id := createExternalVolume(
+			t,
+			append(append(append(append(s3StorageLocations, gcsStorageLocationsNoEncryption...), azureStorageLocations...), s3StorageLocationsNoEncryption...), gcsStorageLocations...),
```
Comment (jdoldis) [inline pkg/sdk/testint/external_volumes_gen_integration_test.go:504]: Once [version 1.22](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/go.mod#L3) is available [Slices.Concat](https://pkg.go.dev/slices#Concat) can be used, which will clean these lines up a bit

File: pkg/sdk/testint/external_volumes_gen_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInt_ExternalVolumes(t *testing.T) {
+	client := testClient(t)
+	ctx := testContext(t)
+
+	awsBaseUrl := testenvs.GetOrSkipTest(t, testenvs.AwsExternalBucketUrl)
```
Comment (jdoldis) [inline pkg/sdk/testint/external_volumes_gen_integration_test.go:20]: Not exactly sure what these evaluate to on your side. The formats I was using locally - 
```
TEST_SF_TF_AWS_EXTERNAL_BUCKET_URL=s3://my_example_bucket
TEST_SF_TF_AWS_EXTERNAL_KEY_ID=1234abcd-12ab-34cd-56ef-1234567890ab
TEST_SF_TF_AWS_EXTERNAL_ROLE_ARN=arn:aws:iam::123456789012:role/myrole
TEST_SF_TF_GCS_EXTERNAL_BUCKET_URL=gcs://my_example_bucket
TEST_SF_TF_AZURE_EXTERNAL_BUCKET_URL=azure://123456789.blob.core.windows.net/my_example_container
TEST_SF_TF_AZURE_EXTERNAL_TENANT_ID=123456789
```

File: pkg/sdk/external_volumes_def.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			IfNotExists().
+			Name().
+			ListAssignment("STORAGE_LOCATIONS", "ExternalVolumeStorageLocation", g.ParameterOptions().Parentheses().Required()).
+			OptionalBooleanAssignment("ALLOW_WRITES", nil).
+			OptionalComment().
+			WithValidation(g.ConflictingFields, "OrReplace", "IfNotExists").
+			WithValidation(g.ValidIdentifier, "name"),
+		storageLocationDef,
+	).
+	AlterOperation(
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/external_volumes_def.go:93]: General question: I see that there are no ALTER UNSET statements inside the docs. Did you try running them manually? I am asking because they are essential for the proper terraform resource design:
- If they are missing, we know we have to have workarounds inside the resource, and we will approach the team to ask for the addition
- If they are just not documented, we can reach out to the doc team and ask for a change.

File: pkg/sdk/external_volumes_def.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+			SQL("EXTERNAL VOLUME").
+			IfExists().
+			Name().
+			WithValidation(g.ValidIdentifier, "name"),
+	).
+	DescribeOperation(
+		g.DescriptionMappingKindSlice,
+		"https://docs.snowflake.com/en/sql-reference/sql/desc-external-volume",
+		g.DbStruct("externalVolumeDescRow").
+			Field("parent_property", "string").
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/external_volumes_def.go:129]: nit-pick: we have helper methods for types, check e.g. https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/pkg/sdk/cortex_search_services_def.go#L62

File: pkg/sdk/external_volumes_gen_test.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+			name: id,
+		}
+	}
+
+	t.Run("validation: nil options", func(t *testing.T) {
+		var opts *AlterExternalVolumeOptions = nil
+		assertOptsInvalidJoinedErrors(t, opts, ErrNilOptions)
+	})
+
+	t.Run("validation: exactly one field from [opts.RemoveStorageLocation opts.Set opts.AddStorageLocation] should be present - two set", func(t *testing.T) {
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/external_volumes_gen_test.go:198]: Please add a test for zero set

File: pkg/sdk/testint/external_volumes_gen_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+		req := sdk.NewCreateExternalVolumeRequest(id, storageLocations).
+			WithIfNotExists(true).
+			WithAllowWrites(allowWrites).
+			WithComment(comment)
+
+		err := client.ExternalVolumes.Create(ctx, req)
+		require.NoError(t, err)
+
+		t.Cleanup(func() {
+			err := client.ExternalVolumes.Drop(ctx, sdk.NewDropExternalVolumeRequest(id))
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/external_volumes_gen_integration_test.go:252]: Please add `if exists` here - usually it does not matter but in a few setups it helps (e.g. test where you want to prematurely remove the object during the test run - with if exists the initially initiated cleanup won't result in an error; e.g. we have rename tests that benefit from it)

File: pkg/sdk/testint/external_volumes_gen_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+	type AzureStorageLocationTrimmed struct {
+		Name            string `json:"NAME"`
+		StorageProvider string `json:"STORAGE_PROVIDER"`
+		StorageBaseUrl  string `json:"STORAGE_BASE_URL"`
+		AzureTenantId   string `json:"AZURE_TENANT_ID"`
+	}
+
+	// Enforce only property names and values in tests, not parent_property, type and property_default
+	// In addition the storage location properties are trimmed to only contain values that we set
+	trimProperties := func(t *testing.T, props []sdk.ExternalVolumeProperty) []ExternalVolumePropNameValue {
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/external_volumes_gen_integration_test.go:105]: Nice function! :)

I think, that ultimately it will have to be added to the production code of the SDK, so that we need an easier access to these values from the terraform resource level. If it happens, it will also require a few unit tests to ensure it handles all the corner case correctly. It will also need a few modifications (like removing the panics) first.

But I would leave it for now as is, but keeping it in mind. cc: @sfc-gh-jcieslak

File: pkg/sdk/testint/external_volumes_gen_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+	awsKmsKeyId := testenvs.GetOrSkipTest(t, testenvs.AwsExternalKeyId)
+	awsExternalId := "123456789"
+
+	gcsBaseUrl := testenvs.GetOrSkipTest(t, testenvs.GcsExternalBuckerUrl)
+	gcsKmsKeyId := "123456789"
+
+	azureBaseUrl := testenvs.GetOrSkipTest(t, testenvs.AzureExternalBucketUrl)
+	azureTenantId := testenvs.GetOrSkipTest(t, testenvs.AzureExternalTenantId)
+
+	assertExternalVolumeShowResult := func(t *testing.T, s *sdk.ExternalVolume, name sdk.AccountObjectIdentifier, allowWrites bool, comment string) {
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/external_volumes_gen_integration_test.go:31]: nit-pick: we try to use the newly generated object assertions (even in the integration tests). You can check e.g. `users_integration_tests`.

Let's leave it like this for now, we will generate these assertions later on while working on the resource

File: pkg/sdk/testint/external_volumes_gen_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+
+		return externalVolumePropNameValue
+	}
+
+	// Storage location structs for testing
+	// Note cannot test awsgov on non-gov Snowflake deployments
+
+	s3StorageLocations := []sdk.ExternalVolumeStorageLocation{
+		{
+			S3StorageLocationParams: &sdk.S3StorageLocationParams{
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/external_volumes_gen_integration_test.go:175]: nit pick: it would be nice to use here the `NewExternalVolumeStorageLocationRequest` (and other builders). The reason: ultimately, we will push every object to be created using these request builders, and this will be the API of the SDK. By using them inside the tests we ensure they are working correctly and that the end users (which includes us on the terraform resource side) can use them correctly.

File: pkg/sdk/external_volumes_def.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	Assignment("STORAGE_PROVIDER", g.KindOfT[S3StorageProvider](), g.ParameterOptions().SingleQuotes().Required()).
+	TextAssignment("STORAGE_AWS_ROLE_ARN", g.ParameterOptions().SingleQuotes().Required()).
+	TextAssignment("STORAGE_BASE_URL", g.ParameterOptions().SingleQuotes().Required()).
+	OptionalTextAssignment("STORAGE_AWS_EXTERNAL_ID", g.ParameterOptions().SingleQuotes()).
+	QueryStructField(
+		"Encryption",
+		g.NewQueryStruct("ExternalVolumeS3Encryption").
+			Assignment("TYPE", g.KindOfT[S3EncryptionType](), g.ParameterOptions().SingleQuotes().Required()).
+			OptionalTextAssignment("KMS_KEY_ID", g.ParameterOptions().SingleQuotes()),
+		g.ListOptions().Parentheses().NoComma().SQL("ENCRYPTION =").Required(),
```
[the remark this answers is not in the document]
Comment (jdoldis) [reply pkg/sdk/external_volumes_def.go]: Great catch, it's not required, have fixed that up

File: pkg/sdk/external_volumes_def.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+		storageLocationDef,
+	).
+	AlterOperation(
+		"https://docs.snowflake.com/en/sql-reference/sql/alter-external-volume",
+		g.NewQueryStruct("AlterExternalVolume").
+			Alter().
+			SQL("EXTERNAL VOLUME").
+			IfExists().
+			Name().
+			OptionalTextAssignment("REMOVE STORAGE_LOCATION", g.ParameterOptions().SingleQuotes().NoEquals()).
```
[the remark this answers is not in the document]
Comment (jdoldis) [reply pkg/sdk/external_volumes_def.go:101]: I think this should already be covered [here](https://github.com/jdoldis/terraform-provider-snowflake/blob/main/pkg/sdk/external_volumes_def.go#L106)

File: pkg/sdk/external_volumes_def.go (hunk 3 above)
Comment (jdoldis) [reply pkg/sdk/external_volumes_def.go:94]: Just tried running `alter external volume test_external_volume_new unset comment;` and got error `Unsupported feature 'UNSET for external volume parameters'.`.

File: pkg/sdk/testint/external_volumes_gen_integration_test.go (hunk 9 above)
Comment (jdoldis) [reply pkg/sdk/testint/external_volumes_gen_integration_test.go:175]: These are all eventually fed into these builders I think, just through the [createExternalVolume](https://github.com/jdoldis/terraform-provider-snowflake/blob/main/pkg/sdk/testint/external_volumes_gen_integration_test.go#L264) function.

**Comment by jdoldis** [comment]
Thanks for the review @sfc-gh-asawicki ! Have pushed changes for most of those suggestions and left a few of comments on others.

File: pkg/sdk/testint/external_volumes_gen_integration_test.go (hunk 9 above)
Comment (jdoldis) [reply pkg/sdk/testint/external_volumes_gen_integration_test.go:175]: For creating not sure how we could use a builder as NewCreateExternalVolumeRequest takes [StorageLocations []ExternalVolumeStorageLocation](https://github.com/jdoldis/terraform-provider-snowflake/blob/main/pkg/sdk/external_volumes_dto_builders_gen.go#L9) directly as a parameter. For altering as you mentioned there is a `NewExternalVolumeStorageLocationRequest`, is being used [in the tests](https://github.com/jdoldis/terraform-provider-snowflake/blob/main/pkg/sdk/testint/external_volumes_gen_integration_test.go#L486) where possible. Let me know if I misunderstood that 👍

File: pkg/sdk/external_volumes_def.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/external_volumes_def.go:101]: Ah yeah, missed that, thanks!

File: pkg/sdk/external_volumes_def.go (hunk 3 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/external_volumes_def.go:94]: ah, that's what I was afraid of. Thanks for checking, we will reach out to the proper team internally and ask for the addition.

**Review by sfc-gh-jcieslak — COMMENTED** [review]
1 comment, otherwise 👍 
Thanks for the contribution :)

File: pkg/sdk/testint/external_volumes_gen_integration_test.go (hunk 9 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/external_volumes_gen_integration_test.go:175]: What I mean was that instead of:

```
s3StorageLocations := []sdk.ExternalVolumeStorageLocation{
	{
		S3StorageLocationParams: &sdk.S3StorageLocationParams{
                         Name:                 "s3_testing_storage_location",
			StorageProvider:      sdk.S3StorageProviderS3,
			StorageAwsRoleArn:    awsRoleARN,
			StorageBaseUrl:       awsBaseUrl,
			StorageAwsExternalId: sdk.String(awsExternalId),
			Encryption: sdk.ExternalVolumeS3Encryption{
				Type:     sdk.S3EncryptionTypeSseKms,
				KmsKeyId: &awsKmsKeyId,
			},
		},
	},
}
```
you can use the request builders too:
```
s3StorageLocations := []sdk.ExternalVolumeStorageLocation{
       NewExternalVolumeStorageLocationRequest().WithS3StorageLocationParams(
                "s3_testing_storage_location",
                 sdk.S3StorageProviderS3,
                 awsRoleARN,
                 awsBaseUrl
        ).
                 WithStorageAwsExternalId(awsExternalId).
                 WithEncryption(
                          NewExternalVolumeS3EncryptionRequest(sdk.S3EncryptionTypeSseKms).
                                   WithKmsKeyId(awsKmsKeyId)
                 ),
}
```

But, to do that, you would have to:
1. change the create operation def for storage locations to `ListQueryStructField("StorageLocations", storageLocationDef, g.ParameterOptions().Parentheses().Required()).`
2. modify the impl file because of this current limitation: https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/65b64d78f94fdc444c482bc52c7e84c5db5b4d2e/pkg/sdk/poc/README.md?plain=1#L135 

For me, it's fine if we leave it as is for now.

**Comment by jdoldis** [comment]
Thanks @sfc-gh-jcieslak , have updated that 🙂

File: pkg/sdk/testint/external_volumes_gen_integration_test.go (hunk 9 above)
Comment (jdoldis) [reply pkg/sdk/testint/external_volumes_gen_integration_test.go:175]: Makes sense, if you're happy to leave that I am too, but if you think it makes sense to implement now I can do that 👍

**Review by sfc-gh-jcieslak — APPROVED** [review]

**Review by sfc-gh-asawicki — APPROVED** [review]
